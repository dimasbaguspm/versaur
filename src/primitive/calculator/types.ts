import type { ButtonHTMLAttributes } from 'react'

/**
 * Props for Calculator component
 */
export interface CalculatorProps {
  /**
   * Initial value to display in the calculator input
   */
  initialValue?: string | number
  /**
   * Callback when the value changes (on equals or input)
   */
  onChange?: (value: string) => void
  /**
   * If true, disables all calculator input
   */
  disabled?: boolean
  /**
   * Optional className for the calculator root
   */
  className?: string
  /**
   * Optional aria-label for accessibility
   */
  'aria-label'?: string
}

export interface CalculatorButtonProps
  extends ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: 'default' | 'action' | 'operator' | 'danger'
}
