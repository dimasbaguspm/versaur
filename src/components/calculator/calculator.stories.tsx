/**
 * Calculator stories for Versaur
 *
 * - Demonstrates usage in modal and with text-input alignment
 * - Shows disabled and initial value states
 */
import { Calculator } from './calculator'
import type { Meta, StoryObj } from '@storybook/react'
import { action } from '@storybook/addon-actions'

const meta: Meta<typeof Calculator> = {
  title: 'Components/Calculator',
  component: Calculator,
  parameters: {
    layout: 'centered',
    docs: {
      description: {
        component: 'A compact calculator for modals and text-input alignment.',
      },
    },
  },
} satisfies Meta<typeof Calculator>

export default meta

type Story = StoryObj<typeof Calculator>

export const Default: Story = {
  args: {
    initialValue: '',
    onChange: action('onChange'),
  },
}

export const WithInitialValue: Story = {
  args: {
    initialValue: '42',
  },
}

export const Disabled: Story = {
  args: {
    disabled: true,
    initialValue: '123',
  },
}
