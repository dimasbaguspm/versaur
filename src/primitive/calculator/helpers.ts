import { cva } from '@/utils/variants'

export const calculatorRootVariants = cva(
  'flex flex-col w-full max-w-xs rounded-lg bg-background',
  {
    variants: {
      disabled: {
        true: 'opacity-60 pointer-events-none',
        false: '',
      },
    },
    defaultVariants: {
      disabled: false,
    },
  }
)

export const calculatorButtonVariants = cva(
  'flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none',
  {
    variants: {
      variant: {
        default:
          'bg-neutral text-foreground hover:bg-neutral-soft hover:text-neutral',
        action:
          'bg-primary text-white hover:bg-primary-soft hover:text-primary',
        operator:
          'bg-secondary text-white hover:bg-secondary-soft hover:text-secondary',
        danger: 'bg-danger text-white hover:bg-danger-soft hover:text-danger',
      },
      disabled: {
        true: 'opacity-50 pointer-events-none',
        false: '',
      },
    },
    defaultVariants: {
      variant: 'default',
      disabled: false,
    },
  }
)
