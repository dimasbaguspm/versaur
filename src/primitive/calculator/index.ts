export { Calculator } from './calculator'
export type * from './types'
