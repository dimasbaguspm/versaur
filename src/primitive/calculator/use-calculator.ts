import { useCallback, useEffect, useRef, useState } from 'react'

export interface UseCalculatorOptions {
  initialValue?: string | number
  disabled?: boolean
  onChange?: (value: string) => void
}

export function useCalculator({
  initialValue = '',
  disabled,
  onChange,
}: UseCalculatorOptions) {
  const [input, setInput] = useState<string>(String(initialValue))
  const inputRef = useRef<HTMLInputElement>(null)

  useEffect(() => {
    setInput(String(initialValue))
  }, [initialValue])

  const evaluate = useCallback((expr: string): string => {
    try {
      if (!/^[-+*/.\d\s]+$/.test(expr)) return ''

      const result = eval(expr)
      return result?.toString() ?? ''
    } catch {
      return ''
    }
  }, [])

  const handleButton = useCallback(
    (val: string) => {
      if (disabled) return
      if (val === 'C') {
        setInput('')
      } else if (val === '⌫') {
        setInput(prev => {
          const next = prev.slice(0, -1)

          return next
        })
      } else if (val === '=') {
        const result = evaluate(input)
        setInput(result)
      } else {
        setInput(prev => {
          const next = prev + val
          return next
        })
      }
    },
    [disabled, input, evaluate]
  )

  useEffect(() => {
    if (onChange) {
      const result = evaluate(input)
      onChange(result)
    }
  }, [evaluate, input, onChange])

  // Keyboard interaction for all keypads
  useEffect(() => {
    if (disabled) return
    const handleKeyDown = (e: KeyboardEvent) => {
      if (
        document.activeElement !== inputRef.current &&
        document.activeElement?.tagName !== 'BODY'
      )
        return
      const key = e.key
      if (key === 'Enter' || key === '=') {
        handleButton('=')
        e.preventDefault()
      } else if (key === 'Backspace') {
        handleButton('⌫')
        e.preventDefault()
      } else if (key === 'Escape' || key === 'C' || key === 'c') {
        handleButton('C')
        e.preventDefault()
      } else if (/^[0-9]$/.test(key)) {
        handleButton(key)
        e.preventDefault()
      } else if (['/', '*', '-', '+', '.'].includes(key)) {
        handleButton(key)
        e.preventDefault()
      }
    }
    window.addEventListener('keydown', handleKeyDown)
    return () => window.removeEventListener('keydown', handleKeyDown)
  }, [handleButton, disabled])

  // Focus input on mount for accessibility
  useEffect(() => {
    if (!disabled) inputRef.current?.focus()
  }, [disabled])

  const handleInput = (e: React.ChangeEvent<HTMLInputElement>) => {
    const val = e.target.value.replace(/[^\d+\-*/.]/g, '')
    setInput(val)
  }

  return {
    input,
    inputRef,
    handleButton,
    handleInput,
  }
}
