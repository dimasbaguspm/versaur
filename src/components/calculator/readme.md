# Calculator Component

A compact, accessible calculator for Versaur UI, designed for use in modals and closely aligned with text-input styling. Supports basic arithmetic, keyboard input, and mobile-friendly layout.

## Features
- Basic arithmetic operations (+, -, *, /)
- Responsive, compact layout
- Keyboard and button input (numbers, operators, Enter, Backspace, Escape, C)
- Always-visible "Done" button for submitting value
- Accessible and screen reader friendly
- Designed to match text-input and modal usage

## Usage
```tsx
import { Calculator } from '@/components/calculator'

<Calculator initialValue="123" onChange={val => console.log(val)} />
```

### Keyboard Interaction
- Numbers, operators, Enter/=/Backspace/Escape/C all work as expected
- "Done" button is not triggered by keyboard, only by click/tap

### Accessibility
- All buttons and input have appropriate aria-labels
- Input is focused on mount for fast keyboard entry
- Fully navigable and operable by keyboard and screen reader

## API
- `initialValue` (string | number): Initial value for the calculator
- `onChange` (function): Called with the value when calculated or Done is pressed
- `disabled` (boolean): Disables all input and buttons
- `className` (string): Custom class for the root
- `aria-label` (string): Custom aria-label for the calculator region
