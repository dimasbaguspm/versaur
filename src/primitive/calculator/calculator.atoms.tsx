import { forwardRef } from 'react'
import { calculatorButtonVariants } from './helpers'
import type { CalculatorButtonProps } from './types'
import { cn } from '@/utils'

export const CalculatorButton = forwardRef<
  HTMLButtonElement,
  CalculatorButtonProps
>(({ variant = 'default', className, ...props }, ref) => (
  <button
    ref={ref}
    type='button'
    className={cn(
      calculatorButtonVariants({ variant, disabled: props.disabled }),
      className
    )}
    {...props}
  />
))
