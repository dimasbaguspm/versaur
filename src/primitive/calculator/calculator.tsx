import { forwardRef } from 'react'
import { CalculatorButton } from './calculator.atoms'
import { calculatorRootVariants } from './helpers'
import type { CalculatorProps } from './types'
import { cn } from '@/utils/cn'
import { ButtonIcon } from '@/primitive/button-icon'
import { X as BackspaceIcon } from 'lucide-react'
import { useCalculator } from './use-calculator'

// Standard calculator layout (no C or ⌫ in the first row)
const BUTTONS = [
  ['7', '8', '9', '/'],
  ['4', '5', '6', '*'],
  ['1', '2', '3', '-'],
  ['0', '.', '=', '+'],
]

export const Calculator = forwardRef<HTMLDivElement, CalculatorProps>(
  (
    {
      initialValue = '',
      onChange,
      disabled,
      className,
      'aria-label': ariaLabel,
    },
    ref
  ) => {
    const { input, inputRef, handleButton, handleInput } = useCalculator({
      initialValue,
      disabled,
      onChange,
    })

    return (
      <div
        ref={ref}
        className={cn(calculatorRootVariants({ disabled }), className)}
        aria-label={ariaLabel || 'Calculator'}
        role='region'
      >
        <input
          ref={inputRef}
          className='w-full mb-3 px-3 py-2 rounded border border-[var(--color-neutral)] bg-[var(--color-neutral-soft)] text-right text-xl font-mono focus:outline-none focus:ring-2 focus:ring-[var(--color-primary)]'
          value={input}
          onChange={handleInput}
          disabled={disabled}
          inputMode='decimal'
          aria-label='Calculator input'
        />
        <div className='flex flex-col gap-1 mb-3'>
          {BUTTONS.map((row, i) => (
            <div key={i} className='flex w-full'>
              {row.map(btn => {
                if (btn === '⌫') {
                  return (
                    <ButtonIcon
                      key='backspace'
                      as={BackspaceIcon}
                      variant='danger-ghost'
                      aria-label='Backspace'
                      size='md'
                      onClick={() => handleButton('⌫')}
                      disabled={disabled}
                    />
                  )
                }
                return (
                  <CalculatorButton
                    key={btn}
                    variant={
                      btn === '='
                        ? 'action'
                        : btn === 'C'
                          ? 'danger'
                          : ['/', '*', '-', '+', '/'].includes(btn)
                            ? 'operator'
                            : 'default'
                    }
                    onClick={() => handleButton(btn)}
                    disabled={disabled}
                    aria-label={btn}
                  >
                    {btn}
                  </CalculatorButton>
                )
              })}
            </div>
          ))}
        </div>
      </div>
    )
  }
)
Calculator.displayName = 'Calculator'
