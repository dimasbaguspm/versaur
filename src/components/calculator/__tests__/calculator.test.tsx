import { render, fireEvent } from '@testing-library/react'
import { composeStories } from '@storybook/react'
import * as stories from '../calculator.stories'

describe('Calculator', () => {
  const { Default } = composeStories(stories)

  it('renders correctly (snapshot)', () => {
    const { asFragment } = render(<Default />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('performs basic addition', () => {
    const { getByText, getByLabelText } = render(<Default />)
    fireEvent.click(getByText('1'))
    fireEvent.click(getByText('+'))
    fireEvent.click(getByText('2'))
    fireEvent.click(getByText('='))
    expect(getByLabelText('Calculator input')).toHaveValue('3')
  })

  // No C or âŒ« in keypad layout, so these tests are removed

  it('supports keyboard interaction for numbers and operators', () => {
    const { getByLabelText } = render(<Default />)
    const input = getByLabelText('Calculator input')
    input.focus()
    fireEvent.keyDown(window, { key: '1' })
    fireEvent.keyDown(window, { key: '+' })
    fireEvent.keyDown(window, { key: '2' })
    fireEvent.keyDown(window, { key: '=' })
    expect(input).toHaveValue('3')
  })

  it('disables input and buttons when disabled', () => {
    const { getByText, getByLabelText } = render(<Default disabled />)
    expect(getByLabelText('Calculator input')).toBeDisabled()
    expect(getByText('1')).toBeDisabled()
    expect(getByText('=')).toBeDisabled()
  })
})
