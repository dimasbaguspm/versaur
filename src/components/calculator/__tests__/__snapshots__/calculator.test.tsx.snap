// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Calculator > renders correctly (snapshot) 1`] = `
<DocumentFragment>
  <div
    aria-label="Calculator"
    class="flex flex-col w-full max-w-xs rounded-lg bg-background p-4"
    role="region"
  >
    <input
      aria-label="Calculator input"
      class="w-full mb-3 px-3 py-2 rounded border border-[var(--color-neutral)] bg-[var(--color-neutral-soft)] text-right text-xl font-mono focus:outline-none focus:ring-2 focus:ring-[var(--color-primary)]"
      inputmode="decimal"
      value=""
    />
    <div
      class="flex flex-col gap-1 mb-3"
    >
      <div
        class="flex w-full"
      >
        <button
          aria-label="7"
          class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-neutral text-foreground hover:bg-neutral-soft hover:text-neutral"
          type="button"
        >
          7
        </button>
        <button
          aria-label="8"
          class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-neutral text-foreground hover:bg-neutral-soft hover:text-neutral"
          type="button"
        >
          8
        </button>
        <button
          aria-label="9"
          class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-neutral text-foreground hover:bg-neutral-soft hover:text-neutral"
          type="button"
        >
          9
        </button>
        <button
          aria-label="/"
          class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-secondary text-white hover:bg-secondary-soft hover:text-secondary"
          type="button"
        >
          /
        </button>
      </div>
      <div
        class="flex w-full"
      >
        <button
          aria-label="4"
          class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-neutral text-foreground hover:bg-neutral-soft hover:text-neutral"
          type="button"
        >
          4
        </button>
        <button
          aria-label="5"
          class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-neutral text-foreground hover:bg-neutral-soft hover:text-neutral"
          type="button"
        >
          5
        </button>
        <button
          aria-label="6"
          class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-neutral text-foreground hover:bg-neutral-soft hover:text-neutral"
          type="button"
        >
          6
        </button>
        <button
          aria-label="*"
          class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-secondary text-white hover:bg-secondary-soft hover:text-secondary"
          type="button"
        >
          *
        </button>
      </div>
      <div
        class="flex w-full"
      >
        <button
          aria-label="1"
          class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-neutral text-foreground hover:bg-neutral-soft hover:text-neutral"
          type="button"
        >
          1
        </button>
        <button
          aria-label="2"
          class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-neutral text-foreground hover:bg-neutral-soft hover:text-neutral"
          type="button"
        >
          2
        </button>
        <button
          aria-label="3"
          class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-neutral text-foreground hover:bg-neutral-soft hover:text-neutral"
          type="button"
        >
          3
        </button>
        <button
          aria-label="-"
          class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-secondary text-white hover:bg-secondary-soft hover:text-secondary"
          type="button"
        >
          -
        </button>
      </div>
      <div
        class="flex w-full"
      >
        <button
          aria-label="0"
          class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-neutral text-foreground hover:bg-neutral-soft hover:text-neutral"
          type="button"
        >
          0
        </button>
        <button
          aria-label="."
          class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-neutral text-foreground hover:bg-neutral-soft hover:text-neutral"
          type="button"
        >
          .
        </button>
        <button
          aria-label="="
          class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-primary text-white hover:bg-primary-soft hover:text-primary"
          type="button"
        >
          =
        </button>
        <button
          aria-label="+"
          class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-secondary text-white hover:bg-secondary-soft hover:text-secondary"
          type="button"
        >
          +
        </button>
      </div>
    </div>
    <button
      aria-disabled="false"
      aria-label="Done"
      class="inline-flex items-center justify-center gap-2 rounded-md font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none bg-coral text-white hover:bg-coral/90 focus-visible:ring-coral-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md h-10 px-4 text-sm min-w-[2.5rem] w-full mt-2"
      type="button"
    >
      Done
    </button>
  </div>
</DocumentFragment>
`;
